### PR TITLE
feat(TableHeader): add id prop

### LINKED ISSUE #10191: [Bug]: TableHeader spreading `...rest` on wrong inner element
### Package

carbon-components-react

### Browser

_No response_

### Package version

7.48.0

### Description

In our code, we had a A11y issue (https://www.w3.org/TR/WCAG20-TECHS/H43) and wanted to make use of the `expandHeader`to help fix this. Going off examples online, I was trying to set the `id` of the th of our table (in thiscase it would be generated via the `TableHeader` component, and so I noticed the source code it has a ...rest prop to pass in any props not explicitly defined. But in our code,. I noticed, only our table headers that were NOT sortable had the id set, but table headers that were sortable did not. Looking at the source code in the `TableHeader`, I noticed this line set the `...rest` in the `th` for columns that are not sortable so that is why the `id` is getting set 

`https://github.com/carbon-design-system/carbon/blob/56cec535e90108731e5a6bb1e20975e4ddd3d1fe/packages/react/src/components/DataTable/TableHeader.js#L74`

However, I noticed for the columns that are sortable, the `th`, does not include the `...rest `, However, I did notice the inner button component DOES include the `...rest`, which makes me thing it was misplaced/typo 

`https://github.com/carbon-design-system/carbon/blob/56cec535e90108731e5a6bb1e20975e4ddd3d1fe/packages/react/src/components/DataTable/TableHeader.js#L116`

So I think this is a bug, (also attached photo to also help)

image

**P.S. I do not mind creating a PR for this fix, I just wish to verify this is a bug first :)** 

### CodeSandbox example

N/A

### Steps to reproduce

I can't reproduce, but provided sources above 

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [X] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

### PR DESCRIPTION
**Pull request by TannerS**
Closes #10191

#### Changelog

**New**

Added a new prop to the TableHeader component to pass in an id to the th element for issue: https://github.com/carbon-design-system/carbon/issues/10191

#### Testing / Reviewing
Added the id to via this new prop to a few story book examples

### CODE REVIEW

File: packages/react/src/components/DataTable/TableHeader.js [hunk 1]
```diff
@@ -141,6 +144,11 @@ TableHeader.propTypes = {
    */
   colSpan: PropTypes.number,
 
+  /**
+   * Supply an id to the th element.
+   */
+  headerId: PropTypes.string,
```
Comment (joshblack) [inline packages/react/src/components/DataTable/TableHeader.js]:
```suggestion
  id: PropTypes.string,
```

Would it make sense to use `id` as the name of the prop or is there something that would not work with that name? 👀

Comment (TannerS) [reply packages/react/src/components/DataTable/TableHeader.js]: Your wish has been granted!

**Review by tay1orjones — APPROVED** [review]
@TannerS looks good 👍  

Tests are failing, but it's an easy fix. Since this is an addition to the public API of this component, the public API snapshot will need to be updated and added to this PR. You can update snapshots by running `yarn test -u`

**Comment by TannerS** [comment]
> @TannerS looks good 👍
> 
> Tests are failing, but it's an easy fix. Since this is an addition to the public API of this component, the public API snapshot will need to be updated and added to this PR. You can update snapshots by running `yarn test -u`

Cool, just ran it

**Review by jnm2377 — APPROVED** [review]

**Pull request merged**